[FLINK-5654] - Add processing time OVER RANGE BETWEEN x PRECEDING aggregation to SQL

rtudoran opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

rtudoran commented:
@fhueske @shaoxuan-wang @twalthr @wuchong @shijinkui 
I re-implemented the time bounded window aggregates in scala. Can you please take a look and merge it. Thanks

hongyuhong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -119,6 +154,64 @@ class DataStreamOverAggregate(
+
+    val result: DataStream[Row] =
+      if (partitionKeys.nonEmpty) {
+        inputDataStreamTimed.keyBy(0)
```
It seems should be inputDataStreamTimed.keyBy(partitionKeys:_*)

rtudoran replied:
Indeed - I missed-saw that ...and forgot to change it 
Thank you

hongyuhong commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala:
```diff
@@ -0,0 +1,104 @@
+     //the design of the Accumulator interface should be extended to enable 
+     //a reset function for better performance
+     while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
```
everytime it should create new accumulator, so i think it is no need to create on open method

rtudoran replied:
indeed. this is more for convenience to have a concise code to access the accumulators

huawei-flink replied:
I will apply this fix also in my function

rtudoran commented:
I have updated the code based on the review. @hongyuhong - thank you very much for the review
@fhueske @twalthr  - can you take a look and check if you can merge it please

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -119,6 +154,64 @@ class DataStreamOverAggregate(
 
   }
 
+  def createTimeBoundedProcessingTimeOverWindow(
```
I am not sure what formatting should be done with respect to this. The method is split in 2 lines in order not to break the 100 characters per line style...
Was done automatically by IDE formater

shijinkui replied:
93 characters, not reach 100. Intellij IDEA can see the result clearly.

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -136,13 +229,13 @@ class DataStreamOverAggregate(
           .name(aggOpName)
           .asInstanceOf[DataStream[Row]]
-        }
+      } // global non-partitioned aggregation
```
I agree...but that was from the code that already existed...from the previous merges.

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -785,7 +785,7 @@ object AggregateUtil {
     (propPos._1, propPos._2)
   }
 
-  private def transformToAggregateFunctions(
```
agree 
@fhueske  - what do you think?

rtudoran replied:
agree 
@fhueske  - what do you think?
I also used the model from other such functions

fhueske replied:
We could also move the logic to create the processing function into `AggreagteUtil` like all other operators that have to deal with aggregations.

rtudoran replied:
@fhueske I could do that - just let me know if this is what you want. For me it would definitely make sense

rtudoran replied:
It is not to much logic...and the overall "createTimeBoundedProcessingTimeOverWindow" is not bigger than the other method for unbound.
...as I said...it is the same...so as you prefer

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateWindowFunction.scala:
```diff
@@ -0,0 +1,108 @@
+  * @param numAggregates The number of aggregates.
+  * @param finalRowArity The arity of the final output row.
+  */
+class DataStreamIncrementalAggregateWindowFunction[W <: Window](
```
my mistake...when I used the previous one as a model.

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateWindowFunction.scala:
```diff
@@ -0,0 +1,108 @@
+  override def open(parameters: Configuration): Unit = {
+     output = new Row(forwardedFieldCount + aggregates.length)
+     accumulators = new Row(aggregates.length)
+     var i = 0
```
in general I agree. but on the other hand it just creates additional objects and overhead. Considering that it is in a very specific case - I prefer to use it like this

fhueske replied:
We can do that in `open()` because it is just called once. However, in the hot path we should use `while` loops because Scala `for` loops have significant overhead.

rtudoran replied:
I will use than everywhere while

rtudoran commented:
@shijinkui  thank you for the remarks. I have addressed them

rtudoran commented:
@fhueske  - can you please take a look over

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -119,6 +154,66 @@ class DataStreamOverAggregate(
+    val lowerboundIndex = index - count
+    var time_boundary = 0L
+     
+    logicWindow.constants.get(lowerboundIndex).getValue2 match {
```
This can be changed to
```
val timeBoundary = logicWindow.constants.get(lowerboundIndex).getValue2 match {
  case _: java.math.BigDecimal => logicWindow.constants.get(lowerboundIndex)
    .getValue2.asInstanceOf[java.math.BigDecimal].longValue()
  case _ => throw new TableException("OVER Window boundaries must be numeric")
}
```

rtudoran replied:
Thanks :)

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -119,6 +154,66 @@ class DataStreamOverAggregate(
+      } else {
+          inputDataStreamTimed.windowAll(GlobalWindows.create()).trigger(CountTrigger.of(1))
+            .evictor(TimeEvictor.of(Time.milliseconds(time_boundary)))        
+            .apply(new DataStreamProcTimeAggregateGlobalWindowFunction[GlobalWindow](
```
I think a `ProcessFunction` would be more efficient for this use case. With a `WindowFunction` and a `GlobalWindow`, we have to aggregate all rows for each emitted row. With a `ProcessFunction` we can just add the new row and retract those rows which are not included anymore.

Moreover, we have to change the logic to a `ProcessFunction` anyway when we extend this to support `FOLLOWING`

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateWindowFunction.scala:
```diff
@@ -0,0 +1,107 @@
+       aggregates(i).resetAccumulator(accumulators.getField(i).asInstanceOf[Accumulator])
+       i += 1
+     }
+     var reuse:Row = null
```
+space

rtudoran replied:
not sure where...

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/DataStreamProcTimeBoundAggIntegrationITCase.scala:
```diff
@@ -0,0 +1,191 @@
+import org.junit._
+import scala.collection.mutable
+
+class DataStreamProcTimeBoundAggIntegrationITCase extends StreamingWithStateTestBase {
```
Please see the comment on PR #3547 regarding the tests.
Thanks

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -17,23 +17,54 @@
 package org.apache.flink.table.plan.nodes.datastream
 
-import org.apache.calcite.plan.{RelOptCluster, RelTraitSet}
+import java.util.{ List => JList }
```
Can you revert the `import` changes?

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -106,6 +137,10 @@ class DataStreamOverAggregate(
         if (overWindow.lowerBound.isUnbounded &&
           overWindow.upperBound.isCurrentRow) {
           createUnboundedAndCurrentRowProcessingTimeOverWindow(inputDS)
+        } else if (overWindow.lowerBound.getOffset.getType.isInstanceOf[IntervalSqlType]
```
@fhueske 
 As a personal preference I like to have the conditions as explicit as possible just to make sure it will not trigger in undesirable cases. I did not knew about the  isRows() function...of course I can check that.
Regarding the upper bound my reasoning was that if we will support other constructs in the future -e.g. FOLLOWING, fixed time boundaries...whatever else will be invented... it is much easier to integrate those if we have this explicit check now..rather then searching for it at that moment
However, if you prefer it removed of course it is not hard to do that :)

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -191,3 +287,31 @@ class DataStreamOverAggregate(
+    override def extractTimestamp(
+      element: Row,
+      previousElementTimestamp: Long): Long = {
+      System.currentTimeMillis()
```
Good point! Thanks.

rtudoran commented:
@fhueske 

"It would also be good if you could remove the merge commit from the PR branch. Merge commits make it much harder to squash commits and merge them to the master branch."

I am sorry but I am not sure what should I do exactly....
Can you give me some pointers of what would you expect?
..The only option I know is to abandon this, recreate a new branch from current master, and move the code and redo the pull request...which I would expect is not the best option...nor what you expect

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -191,3 +287,31 @@ class DataStreamOverAggregate(
 }
 
+object DataStreamProcTimeCase {
+  class ProcTimeTimestampExtractor
```
@fhueske 
Given also your previous comment should I take this completely down?

rtudoran replied:
@fhueske 
Given also your previous comment i would assume this goes down ...which i did

rtudoran commented:
@fhueske - Thank you very much for your review. It was very useful. I integrated and addressed most of the remarks you made.
What remains to be discussed and it is not addressed is:
-using the processingFunction vs keeping the Window-based implementation (i prefer and believe is more appropriate here later as mentioned)
-doing the removal of the merge commit...as per my previous comment i do not  know exactly what to do about that

rtudoran commented:
@fhueske @twalthr  - I am not sure if you saw the updates last Friday with all the things that are going on :)
If you have a chance to look over it would be great - I would like to merge and close this pull reuqest

sunjincheng121 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala:
```diff
@@ -0,0 +1,106 @@
+  * @param aggFields the fields on which to apply the aggregate.
+  * @param forwardedFieldCount The fields to be carried from current row.
+  */
+class DataStreamProcTimeAggregateGlobalWindowFunction[W <: Window](
```
IMHO. we can using `ProcessFunction` implement this feature.

rtudoran replied:
@sunjincheng121 @fhueske 
Thanks for posting. I saw the remark also in the other post. I personally do not believe the processingFunction can be used in this case because of several reasons:
1) if we would use distinct, we cannot extend the implementation - so the design is not extensible/ scalable 
2) Looking at the interface/ example (https://ci.apache.org/projects/flink/flink-docs-release-1.2/dev/stream/process_function.html) i do not see how you can do eviction. We need to take into account the fact that for both JIRA (5653 and 5654) the windows are bounded. That means you need to also eliminate - in one based on time in the other based on number of elements. I am not sure how the processing function can support the eviction (do you have any examples for this)?
3) in case there is not support for eviction that means we need to work with some sort of states in order to keep the elements and manually implement the eviction logic. This would bring a lot of overhead and duplication of code just to not use the windows which directly offer this
4) we need to consider SQL as a whole (IMHO) - which means that we should consider various other parameters that can be added to the aggregation (for example LIMIT x / TOP y... - see https://issues.apache.org/jira/browse/FLINK-6075?filter=-2). This would require that before you apply the logic you have access to the elements to select the ones that are of interest. With processing function you cannot do this. On the other hand building this logic with windows allows you to extend the implementation to support various SQL features

sunjincheng121 replied:
HI, @rtudoran thanks for your explanation. I suggested FLINK-5655(event time OVER RANGE BETWEEN x PRECEDING) jointly consider with this JIRA. not the FLINK-5653. In fact, I working on FLINK-5990(event time OVER ROWS BETWEEN x PRECEDING), After open the FLINK-5990's PR, I'll work on the FLINK-5655,then I can share the example.
Best.
SunJincheng

rtudoran replied:
@sunjincheng121  thanks a lot for the explanation and i am really curious how eviction is supported on the process function so please ping me when you have the PR.

Even assuming this can be done very efficient (and not duplicating the code that exists in windows) the other issues still remain: being generic and extendable for the other SQL functions. For us it is important to support the other parameters of SQL as well...distinct, top, limit... and these require a window in the final end...therefore i believe the best design is the one that can be extended and support more functions. I feel it is also in line with the evolution of the flink project that also enabled to be extended for various scenarios.

fhueske replied:
Actually, I think the `ProcessFunction` is more extensible because the interface is more expressive and allows for more optimizations. We have to manage our own state, but that's rather a good than a bad thing because we have more control over it. That also means that we have to take care of evictions ourselves, i.e., we can have `ListState` and add values or remove them. I agree that it adds more code but not necessarily more overhead. Eviction is not free in DataStream windows, in fact it is rather expensive. Also we would need a custom `Evictor` to remove rows based on processing time.

As I said before, the `ProcessFunction` allows to continuously accumulate and retract values for aggregations. This will be more efficient than aggregating all rows for each row over and over again. Regarding integration of other SQL operators. SQL operators are assembled in a plan. A `LIMIT` or `TOP` operator will be executed after or before an aggregation but not be merged with the same operator. Moreover, a `ProcessFunction` can execute anything and more than a `WindowFunction`.

rtudoran replied:
@fhueske 

I will start with the second part - supporting LIMIT, Distinct, top.... even if these leads to a separate operator that would be applied in theory before we should optimize and combine them...it does not make sense to have 2 operators for one logical operation. Secondly if we refer to keeping them separately - the logic would be that you have a window where you keep all the data to apply limit/top/... and than you fire multiple elements that you selected to the operator to compute the aggregator - so you would reset the aggregator every time.

Regarding your remark of managing the state in process function by ourselves...we are basically creating a window if we keep all elements in the state of the process function. we still need to go through the potentially all elements to evict them and update the aggregator. I am not sure which is more efficiently managed: a large function state or the buffered elements of the window. I would assume in the best case they are similar in performance... if window buffers are not optimized. Nevertheless...there is no advantage as far as i see and understand the problem of having the process function ...but i am sure it brings extra overhead and limitations in the future.

fhueske replied:
A `ProcessFunction` would look like this:

- `open()` In open we create two types of state: 1) the accumulators, which is a `ValueState[Row]`, 2) the rows we need to retract and their processing timestamp. This can be a `ListState[Tuple2[Long, Row]]`.
- `processElement()` is called for each arriving row. We accumulate the row to the accumulator (fetched from the state) and retract all value from the list which fall out of the range (could also be none). We get the current processing time from the Context object. Since the list is sorted on insertion order, we take the head elements. We emit the row with the updated accumulator and append it to the list of rows to retract together with its processing timestamp (again, current time taken from the Context object). Finally we update both states.
- `onTimer()` is not required for processing time OVER window.

If we want to make it more efficient, we could just put those fields into the retraction rows which are actually used by the aggregation function and not the complete row.

What do you think?

sunjincheng121 replied:
@rtudoran I had opened opened the PR #3585 , When I finish the FLINK-5655 I'll ping you again.
@fhueske agree with you. +1
Thanks,
SunJincheng

rtudoran replied:
@fhueske thanks for the design

I will implement it based on the processFunction. The implementation will keep the events as you suggested. For the future implementations we will than use this list state mechanism with the events to do the same operations as we would do them in the table. The 2 are similar in functionality and have the same worst case. Processfunction has a better best case scenario though. It seems that the states and the buffers are similar in processing capabilities - perhaps we should double check it with @aljoscha that there is not performance difference between having a very large state with events and a very large window buffer (e.g. million of events)

aljoscha replied:
There is no difference between a window buffer and state. Internally, when you use an evictor or a `WindowFunction` without an aggregating function (Reduce, Fold, or AggregateFunction) then the window operator will use a `ListState<StreamRecord<T>>` to keep the elements.

rtudoran replied:
@sunjincheng121 @fhueske 
Thanks for the input and suggestion. As of now i followed the model from the Unounded Partiton/noPartition. I have finished acutally the implementation. However, when i run the tests i see that it crashes for the non-partition example
then i checked better and it seems that it also crashes for the:
 UnboundedNonPartitionedProcessingOverProcessFunction.scala
..i do not understand how come

testUnboundNonPartitionedProcessingWindowWithRange(org.apache.flink.table.api.scala.stream.sql.SqlITCase)  Time elapsed: 1.022 sec  <<< ERROR!
org.apache.flink.runtime.client.JobExecutionException: Job execution failed.
	at org.apache.flink.runtime.jobmanager.JobManager$$anonfun$handleMessage$1$$anonfun$applyOrElse$7.apply$mcV$sp(JobManager.scala:915)
	at org.apache.flink.runtime.jobmanager.JobManager$$anonfun$handleMessage$1$$anonfun$applyOrElse$7.apply(JobManager.scala:858)
	at org.apache.flink.runtime.jobmanager.JobManager$$anonfun$handleMessage$1$$anonfun$applyOrElse$7.apply(JobManager.scala:858)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalStateException: Serializer not yet initialized.
	at org.apache.flink.api.common.state.StateDescriptor.getSerializer(StateDescriptor.java:169)
	at org.apache.flink.api.common.state.ListStateDescriptor.getElementSerializer(ListStateDescriptor.java:93)
	at org.apache.flink.runtime.state.DefaultOperatorStateBackend.getOperatorState(DefaultOperatorStateBackend.java:110)
	at org.apache.flink.runtime.state.DefaultOperatorStateBackend.getOperatorState(DefaultOperatorStateBackend.java:91)
	at org.apache.flink.table.runtime.aggregate.UnboundedNonPartitionedProcessingOverProcessFunction.initializeState(UnboundedNonPartitionedProcessingOverProcessFunction.scala:102)
	at org.apache.flink.streaming.api.functions.util.StreamingFunctionUtils.tryRestoreFunction(StreamingFunctionUtils.java:178)
	at org.apache.flink.streaming.api.functions.util.StreamingFunctionUtils.restoreFunctionState(StreamingFunctionUtils.java:160)
	at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.initializeState(AbstractUdfStreamOperator.java:106)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.initializeState(AbstractStreamOperator.java:242)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.initializeOperators(StreamTask.java:681)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.initializeState(StreamTask.java:669)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:251)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:670)
	at java.lang.Thread.run(Thread.java:745)

rtudoran replied:
Ignore my previous comment...stupid error...as it seems that you cannot build and verify just the table library..but need to build the whole project

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamOverAggregate.scala:
```diff
@@ -119,6 +152,60 @@ class DataStreamOverAggregate(
+        inputDS.keyBy(partitionKeys:_*)
+          .window(GlobalWindows.create())
+          .trigger(CountTrigger.of(1))
+          .evictor(TimeEvictor.of(Time.milliseconds(time_boundary))) 
```
@fhueske Actually no - it works on the time that is set. Basically if the user operates on processing time than this is what it will use. It checks at the beginning of the eviction if the record has processing time and if so that it operates on it. I both verified it  manually as well as with tests.

fhueske replied:
No, it does not work. If you remove the line `.assignTimestampsAndWatermarks(new ProcTimeTimestamp())` from the tests (users should not need to assign timestamps for processing time) the `TimeEvictor` does not evict because `TimeEvictor.hasTimestamp()` returns `false`. `TimeEvictor` is only meant for event time and cannot be used for processing time. 

The reason why the test still work is that no element needs to be evicted because the tests execute too fast.

rtudoran replied:
@fhueske i tried also with evicting events and it does work if the .assignTimestampsAndWatermarks is set.
from my point of view if the environment has the ProcessingTime set ...than from my point of view the timestamp should be automatically set to the processing time.
What is your suggestion to handle this issue...we still need to evict based on processing time.

My initial design was to hide this and have this function within the implementation. With that the users do not need to set it and it does not mess up anything ...because in the final end the processing that we build and for which we setup the timestamp is only specific to this query that has the processing time as marker to operate on it.
I would suggest to put it back as it was - what do you think?

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala:
```diff
@@ -0,0 +1,106 @@
+      }
+    }
+
+    //set the values of the result with current elements values if needed
```
@fhueske I would have said that it is always needed to carry on also the contents of the last event...but this is a too strong statement. Although i did not see examples where that is not the case...there are many SQL examples out there. Also, it might be that the optimizer of calcite changes at some point.

fhueske replied:
In Calcite, a LogicalWindow always forwards all fields. Fields which are not needed are removed by a following projection (Calc).

rtudoran replied:
Cool - thanks for the explanation i will update the comment

rtudoran commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateWindowFunction.scala:
```diff
@@ -0,0 +1,108 @@
+       }
+     }
+
+    //set the values of the result with current elements values if needed
```
@fhueske -same as before

rtudoran commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala:
```diff
@@ -317,4 +317,92 @@ class SqlITCase extends StreamingWithStateTestBase {
+      "RANGE BETWEEN INTERVAL '10' SECOND PRECEDING AND CURRENT ROW) as sumC FROM MyTable"
+
+    val t = StreamTestData.get5TupleDataStream(env)
+         .assignTimestampsAndWatermarks(new ProcTimeTimestamp())
```
@fhueske - I agree but Flink as it operates now does not have the timestamp set to the processing time (basically if you check the timestamp of record it is not set if you do not have this set). Therefore you have 2 possibilities: one: to move the assignment of time in the SQL implementation (as it was initially) or to keep it here and expect this is set.
before i had this logic in the processing logic - but your remark was that this should not be set there. Therefore now i moved it here. 
I do not see any alternative as this is quite a nasty case where we do not operate just on time nor just on number of elements.

fhueske replied:
None of the DataStream operators and none of the Table API operators requires to set a processing timestamp. 
If an operator requires this, its design has flaws and should be reconsidered. We should not use the event-time based `TimeEvictor`.

rtudoran replied:
@fhueske - what other option is there to support time boundaries if not time evictor?

rtudoran commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/WindowAggregateTest.scala:
```diff
@@ -239,4 +252,179 @@ class WindowAggregateTest extends TableTestBase {
   }
+  
+   @Test
+  def testMaxAggregatation(): Unit = {
```
@fhueske i have added those as well. i can keep only those

fhueske replied:
Where did you add the tests that check the plan?

rtudoran replied:
I thought you refer to what i added in scala/stream/sql/SqlITCase.scala
...which is the partitioned and non-partitioned verification of 2 aggregates

..i am not sure what you mean with checking the plan....
@fhueske

fhueske replied:
Basically implement tests which are equivalent to the other tests in `WindowAggregateTest`. These tests do not execute a query but compare the plans generated by the optimizer.

rtudoran commented:
@fhueske  
"
I think we cannot use an event-time TimeEvictor to remove records from the GlobalWindow. IMO, we should not automatically overwrite existing timestamps and also not require users to set their own processing-time timestamps. Because of this issue and because a WindowFunction does not support any kind of pre-aggregation, I think that a WindowFunction is not the right approach to implement this type of OVER windows. For example an over window of 1 hour range would aggregate the whole hour for each row over and over again.
"

=> i am not sure there is any other design/option better than this. We could use:
-> processing function - see my comments before
-> we could use window.aggregates (as you suggested to JIRA 5653). There are 2 issues with this: 1) when implementing distinct an the like we still do not have access to the whole data structure ; 2) aggregates is done mainly for supporting just aggregates...which means that in order to support carrying also the contents of the last element (as required by how calcite optimizes the query) you need to artificially create another state-full aggregator to remember the last element (LastElementAggregate) to carry the data - see the solution from @stefanobortoli in 5653 - i feel this is an indication that the design is not the right one either
->we could use windows...

fhueske commented:
PR is continued in #3590
